### ISSUE TITLE
Simplified version as prototype

### ISSUE THREAD

**Issue by jasny** [mindplay-dk/php-generics-rfc#41]
The RFC is too complicated for casual internal contributors to implement. IMHO the ambition is causing this to stay in limbo.

I propose starting with a much-simplified version (even simpler than 'stage 1'). ~~Functionality might be added in the subsequent PHP versions.~~

The following method seems doable;

1. `class Entry<KeyType, ValueType>` would simply register as `Entry` class, removing all `KeyType` and `ValueType` as type hint. This means `new Entry` will just work.
2. Add a `generics` property, which stores the list of generics. Also add 2 hash tables properties to `zend_class_entry`. `functions_generic` store `arg_info` with generic types for each function. `properties_generic` is a properties table with generic types.
3. `new Entry<int,string>` will create a class, similar to an anonymous class. It's named `Entry<int,string>` _(internal class names may use any characters)_ ~~and extends `Entry`~~. For this generic class, the `arg_info` from `functions_generic` and properties from `properties_generic` are copied, replacing `KeyType` and `ValueType` with the keywords `int` and `string`.

One issue is that argument info is currently only used for reflection and not to validate function arguments. I believe there are plans to change this. Otherwise, something similar needs to be added specifically for generics.

**Comment by jasny** [comment]
Default types for `KeyType` and `ValueType` add complexity because `Entry` is no longer a class without type hints for the generic types.

```php
class Entry {
    function foo(int $a) {}
}

class EntryString extends Entry {
    function foo(string $a) {}
}
```

This results in

    PHP Warning:  Declaration of EntryString::foo(string $a) should be compatible with Entry::foo(int $a)

You'd need to implement some specific logic for inheritance that only applies to generics. That seems hard :worried:

**Comment by jasny** [comment]
The `instanceof` logic doesn't seem useful and also complicates things.

```php
if ($cat_box instanceof Box<Feline>) {
   $cat_box->add($feline); // Hmm, that's not right
}
```

With the simplified proposal `Box<Cat>` is simply the class name and that class extends `Box`. No special logic needs to be added for `instanceof`.

**Comment by morrisonlevi** [comment]
Generics are not simple. As an example, I think your proposal is already wrong. Consider this definition:

```php
class Entry<Key, Value> {
    function get(Key $key): Value { /*...*/ }
}
```

You are proposing that the generic type info is elided, so there exists a class like this:

```php
class Entry {
    function get($key) { /*...*/ }
}
```

And then when we instantiate `Entry<int, string>` we get an anonymous class which is equivalent to this:

```php
$obj = new class() extends Entry {
    function get(int $key): string { /*...*/ }
};
```

This violates inheritance rules, because the parent method `Entry::get()` accepted all parameters no matter their type, but in the subclass you can no longer pass any type, only `int`. This would mean you could not pass `$obj` where an `Entry` was expected, which is why it is invalid.

**Comment by jasny** [comment]
Yes, you're right about that. That does bring us to an interesting conclusion though;

If you have a class with method `get(int $key)` and a class with method `get(string $key)`, there is no possibility for them to have a common ancestor that defines method `get`. As such, can `$cat_box` and `$hat_box` actually have the same parent class `Box`? **No**

This allows the implementation to be even a bit simpler, as `Entry<int,string>` doesn't extend `Entry` and is just a copy of `Entry` with different arg info and property types.

**Comment by duesterTechnology** [comment]
Sorry for this question but: Why do you want to inherit from a 'base'-class anyway? This seems to make everything even more complicated. I would agree that a copy would make things much easier than extending from a sort of baseclass

**Comment by mindplay-dk** [comment]
> If you have a class with method get(int $key) and a class with method get(string $key), there is no possibility for them to have a common ancestor that defines method get. As such, can $cat_box and $hat_box actually have the same parent class Box?

[Yes](https://www.typescriptlang.org/play/index.html#code/MYGwhgzhAEBCD2APAPAaWgU0QFwwOwBMY8BXAWwCMMAnAHwm2oEs8BzAGmgDVMd8jo8CgCsMwbAD5oAbwBQ0BdAAOzAG5hc0ANYYAngC5oqANzzFKpus3qQJDIa6mzC4PDwNqJcfGoAKHQZGnDZ2DgCUMs6KCtgAFkwQAHQB0AC82nqm0dFxCYkhGGnQBVmKAL6yUawY2Kh6vmGG6HLZitQ1JNR40LlJAaUKFdFVNVxgthgNDpGtCu3Ynd29+eN2A9AVFbKgkDAAwhoyFTtQ0AAi8KxHlSA10MAaAPoUSEV4GADucEi+AIwATABmTjvL4HbANMJOW7YaAES7PV7pUHfRC+ABEADMmPD0SDPudLpDTEA).

There's more to generics than just type-checking - sometimes you want it just for convenient reuse.

> This allows the implementation to be even a bit simpler, as `Entry<int,string>` doesn't extend `Entry` and is just a copy of `Entry` with different arg info and property types.

What you describe (here and in the original post) is just templates and not really generics.

PHP is a dynamic language - templates aren't really enough: we need to be able to type-check things dynamically, reflect on type-arguments, etc.

I agree that the current proposal has grown very large in scope and complexity - I don't fully understand the need to co/contra variance, for example, and that's where I see the biggest increase in scope right now, relative to my original proposal.

I don't know how to argue (either for or) against it, since I don't fully understand either the feature nor the need.

But I get by in TypeScript (and C# etc.) without this feature just fine, as far as I know - so that would be where I currently see the only meaningful way to reduce the scope of this feature.

Reducing the whole thing to just templates is far too simplistic and, in my opinion, doesn't make sense in a dynamic language - you can get basic static templates [right now](https://psalm.dev/docs/annotating_code/templated_annotations/) without any language extension.

I'm inclined to close this issue.

@morrisonlevi @natebrunette @nikic @orolyn thoughts?

**Comment by duesterTechnology** [comment]
> Reducing the whole thing to just templates is far too simplistic and, in my opinion, doesn't make sense in a dynamic language - you can get basic static templates right now without any language extension.

This is kind of a soft templating. From my point of view, PHP is changing from this kind of annotation-based-type-checks to a real typed language (at least as an optional alternative).  We should start with something simple to get the implementation started and have some kind of base - eventhough it seems too simple for the moment and does not cover the whole generic part.

**Comment by mindplay-dk** [comment]
@duesterTechnology sure, but now you're describing an implementation approach - not the outcome. (not sure if that's actually your point? if so, we don't need an issue to track it.)

**Comment by duesterTechnology** [comment]
Yes, sure, it is an implementation approach. I got it the way that both is part of the discussion here as both correlate with each other. Sorry if I misunderstood that point. But the argument that a templating exists as a 'meta'-feature within comments cannot be an argument against the suggestion. Because from my point of view, the approach discussed here sould be a type-safe solution. If not in its first version, then in the subsequent versions. So I was about to consider both, the implementation side and the outcome. Did I get you right this time?
Edit: The main point in this issue seems to be the discussion if a 'simple' implementation for php8 shounld be done to get started. And I think: yes, we should start with something simple and look how it works. More functionality can be added afterwards.

**Comment by mindplay-dk** [comment]
@duesterTechnology to be clear: I don't think there should be a "first version" of this - I've already explained why. As an implementation approach, of course, I am fine with implementing a subset of the features - but not as an outcome. (and a release is an outcome.)

Honestly, rather than a limited first version, I'd prefer to see an optimized prototype of the full feature set. Even if this is implemented in plain PHP and has really poor performance, a simple prototype allows us to test the concepts, and it's more malleable, allowing easier changes to the concepts and features overall. Building a proper implementation of a subset of the features, even if everyone is happy with those in isolation, once you add the next layer of features, you may discover the first layer needs substantial changes, and it will most likely make it much harder to change anything.

Either way, this is just further musings on implementation approach - at this stage, I'm more interested in a seeing a complete set of requirements than a partial implementation based on half-baked requirements. If we had all the requirements, I'd prefer to see a prototype.

I don't think think I'm really adding anything new to the discussion at this point though, so at this stage I'd prefer to let others comment.

**Comment by mindplay-dk** [comment]
Sorry, I just realized the past comment I referred to was never posted, it was sitting in a tab on another machine 🙄

I'll post those comments here:

> Yes, sure, it is an implementation approach. I got it the way that both is part of the discussion here as both correlate with each other.

What we're discussing at this point is the spec/requirements - so, issues in the issue tracker are issues with the spec, which we're trying to refine and complete.

That doesn't mean we can't talk about implementation, in the context of individual issues and requirement - but the original post says:

> Functionality might be added in the subsequent PHP versions.

This implied to me a request to reduce the scope of the implementation and *release* it, in hopes of expanding on the feature in a later release.

*(below is the main point I was referring to in my comment above)*

History shows this approach is a *very* bad idea for programming languages - it is almost impossible to predict the future requirements of the final feature, and so there is almost always breaking changes introduced while expanding on a feature, unless (much worse) trade-offs and compromises ultimately lead to an inferior or confusing finished feature, unless of course the feature is just left in a half-baked or incomplete state forever.

For almost any project *other* than a programming language, I'm all for releasing early and enhancing features over time - the problem with (interpreted) languages is, once a feature is in use, you can't really change it.

That said, I'd be happy if we divided things into milestones, to guide the implementation along. (I don't know enough about implementing this feature to do that myself, but it sounds like a good idea.)

*(again, sorry, I thought I had already posted these comments!)*

**Comment by duesterTechnology** [comment]
Ah, okay, got you now. Sorry for my missunderstanding. Well, I now agree with your point. A protorype would be a good thing anyway to test things.

**Comment by jasny** [comment]
:+1:  A prototype is fine as compromise. I've changed the title to reflect this.

**Comment by zmitic** [comment]
May I suggest something from end-user perspective and ask not to hate me? 

TL;DR:
**Validate syntax only, not types.**

---
Let me explain why. People who want generics are not beginners; we all fully understand their value, we do typehint our code and we use static analysis tools. Not any of us would return wrong type and not notice it; it just doesn't happen in reality.

Most of the generics would deal with iterables, so I will use this realistic but simplified example:

```php
public function process(): void
{
	foreach($this->getProducts() as $product) {
		$price = $product->getPrice();
		// do something with price
	}
}

private function getProducts(): iterable<Product>
{
	// this would be database results
	yield new Product();
	yield new Product();
}
```

My idea is that if user writes something like this:

```php
private function getProducts(): iterable<Product> 
{
	yield new Product();
	yield new Product();
	yield new User();	// wrong here
}
```

literally **nothing** would happen during execution; I think Hack works the same way. But instead of integrated analysis, we use phpstan/psalm/phan and PHPStorm. 
That is why I said it is impossible not to see the problem.

---
Keeping BC would not be a problem:

```php
// some Symfony controller
public function doSomething(): Response
{
	$user = $this->getUser<User>();
}
```

Real Security class would need minimum changes to keep BC and support generics at the same time:
```php
class Security
{
    /**
	 * @template T
     * 
     * @return T|UserInterface|null
     */
    public function getUser()
	{
		// no changes here
	}
}
```

and this would work; types would be ignored by PHP engine, but tools would not have problems with this.


Doctrine entity:
```php
class Product
{
	private ArrayCollection<Category> $categories;

	public function __construct()
	{
		$this->categories = new ArrayCollection<Category>();
	}
}
```

Again; by simply ignoring types, existing code would work without BC problems.

---
Yes, it does break how PHP works but is it really important for prototype? I doubt because of reasons above. Remember; those who want generics are not beginners and that is my entire point.


Because they are so complicated and didn't move at all, I think validating syntax only is enough for prototype.


If PHP contributors are worried about beginners misusing generics, php.ini could have them disabled by default and would trigger compilation error. 

I understand this in unpopular idea to ignore typehints but I ask you to think about people who would use generics and real use cases. But real, not hypothetical.

---
And one more; because of complexity and limited resources, I would suggest to open a bounty. I am pretty sure that even this basic prototype could collect enough money to motivate contributors.

**Comment by mindplay-dk** [comment]
> Validate syntax only, not types.

@zmitic this is of course fine as a step during prototyping.

As a language feature, this would be quite incomplete and very incoherent with PHP, being a reflected language, and a run-time type-checked language.

Consistency is important - deviations from the overall direction of the language become nasty surprises; the kind of stuff that gives PHP name as an inconsistent, messy language. We don't want to make things worse by adding more of that.

Hack tried this direction, and, in my personal opinion, it added more problems than benefits.

**Comment by zmitic** [comment]
@mindplay-dk 

> As a language feature, this would be quite incomplete and very incoherent with PHP, being a reflected language, and a run-time type-checked language.

I agree but there hasn't been anything for 3 years and no plans to start. Basic prototype, as described above, would at least be something that could bring more brain power.

Also, with php.ini config users would know they are dealing with fragile things not meant for beginners. As stated above; experienced developers are those who want them, not people just learning about typehints.

> Consistency is important - deviations from the overall direction of the language become nasty surprises; the kind of stuff that gives PHP name as an inconsistent, messy language. 

I think the lack of generics gives even worse name.

> Hack tried this direction, and, in my personal opinion, it added more problems than benefits.

Can you elaborate? I haven't read anything about problems but I don't follow it very much; get envy every time :smile:

**Comment by zmitic** [comment]
One more issue; this simplified prototype is maybe even better to implement first then full version (i.e. the one that doesn't ignore types), even if it was possible.

##### The reason
generics are a game changer and it will take time for libraries and frameworks to catch-up. It sounds weird but we wouldn't be able to use them until all libs support them.

Example: if I was to use fictional PHP with full generics, code like this would fail:

```php
$user = $this->getUser<User>();
```

till Symfony changes all core methods.
 
But by ignoring types and relying on external tools, it will give FW developers enough time for major revision.

**Comment by jasny** [comment]
@zmitic Well... I have started, but it's kinda hard. Even just to get started. I've modified the parser so it accepts the syntax for generics.

If you want to follow it, checkout https://github.com/jasny/php-src/tree/generics

**Comment by zmitic** [comment]
@jasny Hope you will make it, I have zero C knowledge. Or someone could help; it is not just to accept the syntax, Reflection classes are also a challenge.
